Allow custom profiling of JS and non-JS allocation counts

ddegazio opened the pull request:
#### 53acbba3ab24d8ee1ddc5f488e7941ea3cf3ebfe
<pre>
Allow custom profiling of JS and non-JS allocation counts
<a href="https://bugs.webkit.org/show_bug.cgi?id=262827">https://bugs.webkit.org/show_bug.cgi?id=262827</a>
rdar://114841891

Reviewed by Mark Lam.

Adds macros to FastMalloc, IsoHeap, and Gigacage malloc functions to allow
for custom profiling code around different kinds of mallocs. Profiling for
different categories of mallocs can be provided by defining certain macros
(see AllocationCounts.h for more information).

* Source/WTF/wtf/Gigacage.cpp:
(Gigacage::tryAlignedMalloc):
(Gigacage::tryMalloc):
(Gigacage::tryZeroedMalloc):
(Gigacage::tryRealloc):
* Source/bmalloc/CMakeLists.txt:
* Source/bmalloc/bmalloc.xcodeproj/project.pbxproj:
* Source/bmalloc/bmalloc/AllocationCounts.cpp: Added.
* Source/bmalloc/bmalloc/AllocationCounts.h: Added.
* Source/bmalloc/bmalloc/BCompiler.h:
* Source/bmalloc/bmalloc/IsoHeap.cpp:
(bmalloc::api::isoAllocate):
(bmalloc::api::isoTryAllocate):
* Source/bmalloc/bmalloc/bmalloc.h:

Canonical link: <a href="https://commits.webkit.org/269033@main">https://commits.webkit.org/269033@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0705db10df7c131450112d959cfe1858a098a50e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/21385 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/21694 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/22437 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/23246 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/19821 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/24993 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21939 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/21027 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/21609 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/21258 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/18529 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/24098 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/18426 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/19378 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/25702 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/18595 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/19507 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/19585 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/23553 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/20762 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20071 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/17089 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/24789 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/19388 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/5862 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/23632 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/26060 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2652 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19976 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5689 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

MenloDorian reviewed (approved):
r=me if EWS is happy.

ddegazio committed:
Allow custom profiling of JS and non-JS allocation counts
https://bugs.webkit.org/show_bug.cgi?id=262827
rdar://114841891

Reviewed by Mark Lam.

Adds macros to FastMalloc, IsoHeap, and Gigacage malloc functions to allow
for custom profiling code around different kinds of mallocs. Profiling for
different categories of mallocs can be provided by defining certain macros
(see AllocationCounts.h for more information).

* Source/WTF/wtf/Gigacage.cpp:
(Gigacage::tryAlignedMalloc):
(Gigacage::tryMalloc):
(Gigacage::tryZeroedMalloc):
(Gigacage::tryRealloc):
* Source/bmalloc/CMakeLists.txt:
* Source/bmalloc/bmalloc.xcodeproj/project.pbxproj:
* Source/bmalloc/bmalloc/AllocationCounts.cpp: Added.
* Source/bmalloc/bmalloc/AllocationCounts.h: Added.
* Source/bmalloc/bmalloc/BCompiler.h:
* Source/bmalloc/bmalloc/IsoHeap.cpp:
(bmalloc::api::isoAllocate):
(bmalloc::api::isoTryAllocate):
* Source/bmalloc/bmalloc/bmalloc.h:

Canonical link: https://commits.webkit.org/269033@main
